### PR TITLE
Added tracing for subscription and commit end points

### PR DESCRIPTION
**Pull request by Kunal-Jha**
# Added tracing for subscription and commit endpoints

## Zalando ticket: ARUHA-2547
https://jira.zalando.net/browse/ARUHA-2547

## Description
This PR adds tracing for the consumption and committing of data to Nakadi. This would give the publisher the visibility of their events when publishing to Nakadi.

### CODE REVIEW

File: src/main/java/org/zalando/nakadi/filters/TracingFilter.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+            GlobalTracer.get().inject(baseSpan.context(),
+                    HTTP_HEADERS, new TextMapInjectAdapter(spanContextToInject));
+            response.setHeader(SPAN_CONTEXT, spanContextToInject.toString());
+            baseSpan.finish();
+        }
+    }
+
+
+    private RequestType getRequestType(final RequestInfo requestLogInfo) {
+        if (requestLogInfo.path != null && "POST".equals(requestLogInfo.method) &&
```
Comment (thomasabraham) [inline src/main/java/org/zalando/nakadi/filters/TracingFilter.java]: Maybe we can add `if(requestLogInfo.path == null){ return RequestType.OTHER; }` above and remove the null check from individual if cases.

File: src/main/java/org/zalando/nakadi/service/TracingService.java [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package org.zalando.nakadi.service;
 
 import com.google.common.collect.ImmutableMap;
+import io.opentracing.References;
```
Comment (thomasabraham) [inline src/main/java/org/zalando/nakadi/service/TracingService.java]: Why `TracingService` is annotated with `@Configuration`?

File: src/main/java/org/zalando/nakadi/filters/TracingFilter.java [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        if (requestType.equals(RequestType.OTHER)) {
+            return;
+        }
+        final Map<String, Object> tags = new HashMap<String, Object>() {{
+            put("http.status_code", statusCode);
+            put("content_length", requestLogInfo.contentLength);
+        }};
+
+        if (requestType.equals(RequestType.PUBLISHING)) {
+            String sloBucket = "5K-50K";
```
Comment (thomasabraham) [inline src/main/java/org/zalando/nakadi/filters/TracingFilter.java]: Is this duplicated in Publishing controller?

File: src/main/java/org/zalando/nakadi/controller/EventPublishingController.java [hunk 4]
```diff
@@ -78,36 +78,34 @@ public ResponseEntity postEvent(@PathVariable final String eventTypeName,
         final HttpServletRequest httpServletRequest = request.getNativeRequest(HttpServletRequest.class);
         final Scope publishingScope = TracingService
                 .activateSpan((Span) httpServletRequest.getAttribute("span"), false);
+        String sloBucket = "5K-50K";
```
Comment (thomasabraham) [inline src/main/java/org/zalando/nakadi/controller/EventPublishingController.java]: Is this somehow duplicate?

**Comment by thomasabraham** [comment]
👍

File: src/main/java/org/zalando/nakadi/controller/CursorsController.java [hunk 5]
```diff
@@ -88,15 +93,16 @@ public CursorsController(final CursorsService cursorsService,
         if (cursors.isEmpty()) {
             throw new CursorsAreEmptyException();
         }
-
+        final Scope commitScope = TracingService.activateSpan((Span) request.getAttribute("span"), false);
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/controller/CursorsController.java]: can we have `TracingService.activateSpan` take `request`, and do all the casting within the method ?

File: src/main/java/org/zalando/nakadi/controller/EventPublishingController.java [hunk 6]
```diff
@@ -78,36 +78,35 @@ public ResponseEntity postEvent(@PathVariable final String eventTypeName,
         final HttpServletRequest httpServletRequest = request.getNativeRequest(HttpServletRequest.class);
         final Scope publishingScope = TracingService
                 .activateSpan((Span) httpServletRequest.getAttribute("span"), false);
+        publishingScope.span().setOperationName("publish_events");
+        String sloBucket = "5K-50K";
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/controller/EventPublishingController.java]: make sense to extract to the method and move values to constants
`getSloBucket(httpServletRequest)`

**Comment by adyach** [comment]
@Kunal-Jha in general I agree with @rcillo that having logic to identify the commit cursor endpoint in filter is not the best way to go, I would rather pass up on have a reference span. At the same time having refactoring the whole exceptions handing logic is also not an option otherwise it is not possible to finish spans with specific errors context which will have a lot more changes in the existing code, so I would say it is fine to go with this changes. Maybe @rcillo has something to comment about here.

**Comment by adyach** [comment]
👍

**Comment by Kunal-Jha** [comment]
👍

**Pull request merged**